### PR TITLE
Make Cabal agnostic about working directory

### PR DESCRIPTION
**Pull request by sheaf**
**Template Α: This PR modifies `cabal` behaviour**

This PR tackles #9702, making the library functions in `Cabal` agnostic to the working directory. In practice, this means that we distinguish `FilePath`s from un-interpreted `SymbolicPath`s. The latter may be paths that are relative to the working directory, and need to be interpreted with respect to a passed-in argument specifying the working directory, instead of using the working directory of the current process.
See Note [Symbolic paths] in Distribution.Utils.Path.

In particular, paths in the package description now use the `SymbolicPath` abstraction, which allows specifying whether they are allowed to be absolute, and, if they are relative, what they are relative to.
For example, source files are relative to a source search directory, data files are relative to the data directory, and doc files are
relative to the package root.

---

  * [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
  * [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [x] The documentation has been updated, if necessary.
  * [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
  * [x] Tests have been added.

TODO:
  * [ ] More extensive testing, e.g. compiling `head.hackage`.

## QA notes

The testing strategy for this patch is simply to build, test & benchmark packages. The main difference is that we no longer set the working directory when invoking the `Cabal` library internally (see `internalSetupMethod`), relying instead on passing `-working-dir`.

### CODE REVIEW

**Comment by sheaf** [comment]
I have compiled [`clc-stackage`](https://github.com/Bodigrim/clc-stackage) using this branch. It showed up a single failure which is unrelated to this patch and that I filed separately as #9777 . 

I now consider this MR ready for review.

File: cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs [hunk 1]
```diff
@@ -262,27 +270,30 @@ buildAndRegisterUnpackedPackage
         | hasValidHaddockTargets pkg = action
         | otherwise = return ()
 
+      mbWorkDir = useWorkingDir $ scriptOptions
```
Comment (geekosaur) [inline cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs:273]: Why the `$`?

**Review by geekosaur — COMMENTED** [review]
TBH I'm not convinced this review is any better than CI + hlint 😀

File: cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs (hunk 1 above)
Comment (philderbeast) [reply cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs:273]: As you've noted @geekosaur. This is one of those things we could be relying on HLint for, instead of checking for things like this manually at review, #9110;

```diff
git diff
diff --git a/.hlint.yaml b/.hlint.yaml
index e38cc7be7..7c9f4cbbd 100644
--- a/.hlint.yaml
+++ b/.hlint.yaml
@@ -13,7 +13,6 @@
 - ignore: {name: "Monoid law, left identity"} # 3 hints
 - ignore: {name: "Monoid law, right identity"} # 3 hints
 - ignore: {name: "Move guards forward"} # 4 hints
-- ignore: {name: "Redundant $"} # 125 hints
 - ignore: {name: "Redundant $!"} # 4 hints
 - ignore: {name: "Redundant <$>"} # 6 hints
 - ignore: {name: "Redundant =="} # 1 hint
```

```
$ hlint cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs
cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs:273:33: Suggestion: Redundant $
Found:
  useWorkingDir $ scriptOptions
Perhaps:
  useWorkingDir scriptOptions

1 hint
```

File: Cabal-syntax/src/Distribution/Types/BuildInfo.hs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE DataKinds #-}
```
[the remark this answers is not in the document]
Comment (sheaf) [reply Cabal-syntax/src/Distribution/Types/BuildInfo.hs:1]: The use of this extension is, I believe, entirely due to changing the design of the symbolic path datatype. It used to be that you declared individual empty data-types:

```haskell
data PackageDir
data SourceDir
data SourceFile
instance IsDir PackageDir
instance IsDir SourceDir
```

and then you would have

```haskell
type SymbolicPath :: Type -> Type -> Type

hsSourceDirs :: [ SymbolicPath PackageDir SourceDir ]
```

I didn't like the typeclass mechanism using `IsDir`; I think it makes a lot more sense to distinguish directly in the type whether something is a file or a directory instead of relying on a typeclass for which instances could get added later on. So I then went for:

```haskell
data FileOrDir a = File a | Dir a
type SymbolicPath :: a -> FileOrDir a -> Type

hsSourceDirs :: [ SymbolicPath PackageDir (Dir SourceDir) ]
```

This gets rid of the instances, but it does use `DataKinds`.

It was still a bit of a chore to define all these empty data-types, so I changed it further to use type-level strings:

```haskell
data FileOrDir = File Symbolc | Dir Symbol
type SymbolicPath :: Symbol -> FileOrDir -> Type

hsSourceDirs :: [ SymbolicPath "Package" (Dir "Source") ]
```

Now, it can be argued that it is better to have datatypes rather than raw strings: a datatype is properly namespaced by the module that defines it, and has to be imported in order to be referred to. This avoids issues in which two different modules define a different type of abstraction location with the same symbolic name. Given the amount of `SymbolicPath`s now in the codebase, it lead to a proliferation of data type declarations that in my opinion bloated the patch. In particular, it made it hard to iterate on the patch as each time I came across a new kind of abstract filepath I had to find in which module it made sense to define a corresponding new datatype, and then had to change import/export lists to include it.

[the remark this answers is not in the document]
Comment (sheaf) [reply Cabal-syntax/src/Distribution/Types/BuildInfo.hs:1]: I'm also using `DataKinds` to distinguish symbolic paths that must be relative from symbolic paths that are allowed to be absolute:

```haskell
-- | Is this symbolic path allowed to be absolute, or must it be relative?
data AllowAbsolute
  = -- | The path may be absolute, or it may be relative.
    AllowAbsolute
  | -- | The path must be relative.
    OnlyRelative
```

This allows functions to be polymorphic about whether paths must be absolute or relative, for example:

```haskell
findFileCwd
  :: forall dir1 dir2 file allowAbsolute
   . Verbosity
  -> Maybe (SymbolicPath "CWD" (Dir dir1))
  -- ^ working directory
  -> [SymbolicPathX allowAbsolute dir1 (Dir dir2)]
  -- ^ search directories
  -> RelativePath dir2 (File file)
  -- ^ File Name
  -> IO (SymbolicPathX allowAbsolute dir1 (File file))
  ```
  
If `findFileCwd` gets passed in search directories that are guaranteed to be relative, it will return file paths that are themselves relative. Without `DataKinds`, I think I would end up having to define two separate datatypes and then a function such as this would have to inherit a class constraint in order to retain polymorphism.

[the remark this answers is not in the document]
Comment (augustss) [reply Cabal-syntax/src/Distribution/Types/BuildInfo.hs:1]: Don't worry about.  I'll write my own cabal replacement.  The source for cabal is too far gone. :)

**Comment by alt-romes** [comment]
I think this patch is a very good improvement to the cabal codebase, and I believe in its correctness additionally because of the robust testing strategy.

I think it is ready.

@andreabedini would you care to comment/review this patch, so we can proceed swiftly to merging before it starts bitrotting?

**Comment by geekosaur** [comment]
It might make sense to split it into multiple reviewers by chunks, since it's an awfully big gulp for a single reviewer.

**Comment by sheaf** [comment]
> It might make sense to split it into multiple reviewers by chunks, since it's an awfully big gulp for a single reviewer.

I understand why you say that, but the way I think of this patch is that it boilds down to changing every direct interaction with the filesystem (e.g. `doesDirectoryExist`, `createFile` etc) to account for a working directory argument. I don't think it's possible to audit every such change, which is why I focused on testing the patch extensively. Instead, I think a review should focus on the overall new design and direction of travel, without necessarily looking at every changed file.

**Review by fendor — APPROVED** [review]
I reviewed the main Path.hs module in detail and sporadically looked over the changed files.

The explicit handling of the working directory looks like a clear improvement to me and I like it a lot!
It allows me to see at a glance what kind of paths I will need for a function, and it is more type safe. +1 from me to get this merged asap.

I see two main disadvantages of the proposed approach:
* No pattern matching for filepaths
  * Not an issue, imo, as we should not be using that anyway for any non-trivial pattern
* More complex language features
  * Also no big deal if they are well-documented, which they are in this case.
  
I found one or two confusing code patterns, I think it might be nice to explain a couple of these to give onboarding devs an intuitive understanding of the `SymbolicPath` infrastructure.
 For example:
 
* `RelativePath "Source" (File "Source")`, at a glance, reads as relative to itself, although it is likely a source file in the "Source" location? Might be nice to add a short explanation in a note.

At last, we should add a naming convention to the location names, e.g. `Source`, `Package`, `CWD`. Perhaps it is also interesting to newcomers that the casing matters. 

Further, the most common locations should have a short explanation for what they represent.   For example: 

> "Package" is the location where a `.cabal` file is found. 

A note in `Path.hs` would be fine by me.

All in all, awesome improvement, can't wait to have this while working on a cabal patch!

**Comment by andreabedini** [comment]
Sorry for the extended radio silence. I am not going to read this line by line but I want to commend @sheaf for the great and detailed work.

I have to admit I feel slightly dissatisfied by `AllowAbsolute`, I feel the information on whether the path is (or can be) absolute should go along with `from :: Symbol`. I do not understand what `Dir "Package"` tells me in presence of `AllowAbsolute`?

Does the type `SymbolicPathX AllowAbsolute (Dir "Package") (File "Source")` mean that the file path can be either:
- relative to _"Package"_, e.g. `src/Utils/Path.hs`
- absolute *and* extending inside _"Package"_ e.g `/home/user/workdir/cabal/Cabal-syntax/src/Utils/Path.hs`
?

If I have
```
sp = SymbolicPath "/some/directory/Cabal-syntax/Cabal-syntax.cabal" :: SymbolicPathX AllowAbsolute (Dir "Package") (File ".cabal")`
```
then (for all `to`)
```
takeDirectory sp :: SymbolicPathX AllowAbsolute (Dir "Package") (Dir to)
```

Maybe it could be `Dir Unknown` instead? (We could consider that the cabal file is always in the package root directory but that's another story).

Is there any chance we end up going past the package directory and produce something like the following?
```
SymbolicPath "/some/directory" :: SymbolicPathX AllowAbsolute (Dir "Package") (Dir _)`
```

Semantically speaking `AllowAbsolute` subsumes `Reative` so I would find more natural to use polymorphism, i.e.

```
newtype SymbolicPath (root :: k) (to :: FileOrDir) = SymbolicPath FilePath
data Absolute
data RelativeTo (what :: Symbol)

SymbolicPath "src/Utils/Path.hs" :: SymbolicPath (RelativeTo "Package") (File "Source")
```
This would be similar to the design in `hackage-security`.

**Comment by sheaf** [comment]
> This would be similar to the design in `hackage-security`.

@andreabedini Thanks for the pointer, I wasn't aware that `hackage-security` [vendored its own `Path` abstraction](https://github.com/haskell/hackage-security/blob/f18ad437bcdf16e7b40e0e1785948cfcda42629c/hackage-security/src/Hackage/Security/TUF/Paths.hs).

As for your comment about keeping track of whether a path is absolute or relative in the type: the most common use-case within Cabal is of a path that is either absolute or relative to a particular directory. For example, source files generated by preprocessors end up in the build directory, which can well be outside of the current package directory; see e.g. [`preprocessFile`](https://github.com/haskell/cabal/blob/1c1230ca228cc03a9ee68166243af358ab3992fc/Cabal/src/Distribution/Simple/PreProcess.hs#L300-L304). This means that, for example, Haskell source search directories are "either absolute or relative to the package".
I opted for a design that reduced friction for this particular-use case; a design that kept track of whether a path is absolute or relative in the type would introduce a lot of existentials, wrapping/unwrapping, RankNTypes, and so on.

**Comment by sheaf** [comment]
To reiterate, the concept that seems most useful, at least for the purposes of this working directory change, is the notion of a filepath which, if it is relative, is relative to the package. This is what the `SymbolicPath from to` type definition provides.

An existential such as

```haskell
data MyPath to where
  RelPath :: RelativePath from to -> MyPath to
  AbsPath :: AbsolutePath to -> MyPath to
```

loses this information: if I have a `MyPath to` in hand, and it ends up being a relative path, I have forgotten what it was relative to.

**Comment by sheaf** [comment]
I have updated the SymbolicPath API thusly:

  - we no longer distinguish different kinds of files (in practice it didn't seem to catch any bugs, and made some things a bit tedious),
  - directories are back to using datatypes instead of `Symbol`,
  - I added a `interpretSymbolicPathCWD` function that should be used instead of `getSymbolicPath`;
    it has the same body as `getSymbolicPath` but imposes an additional constraint that should be discharged.
    See `Note [Working directory proof tokens]`. In practice, this helps ensure we don't accidentally discard
    a working directory argument.
    
I believe the first two changes make the API easier to use and should address @fendor's review. It means we no longer need a naming convention or worry about capitalisation, and the individual abstract directories now have appropriate Haddock documentation.

I have not made any changes involving `AllowAbsolute`, as I believe the current design is more appropriate for the needs of `Cabal`.

I am currently re-building stackage as a safety check.

**Review by andreabedini — APPROVED** [review]

File: Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs [hunk 3]
```diff
@@ -46,7 +47,7 @@ licenseRaw :: Lens' PackageDescription (Either SPDX.License License)
 licenseRaw f s = fmap (\x -> s{T.licenseRaw = x}) (f (T.licenseRaw s))
 {-# INLINE licenseRaw #-}
 
-licenseFiles :: Lens' PackageDescription [SymbolicPath PackageDir LicenseFile]
+licenseFiles :: Lens' PackageDescription [RelativePath Pkg File]
```
Comment (phadej) [inline Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs:50]: I'm very unhappy to see these kind of changes. You break literally all `Cabal-syntax` consumers, and you don't get any benefit. The `SymbolicPath PackageDir LicenseFile` was already *relative*, it's a (relative) path from package directory to the license file.

You also drop some type-safety, as `LicenseFile` was telling that it's not *some* file, but a *license* file (e.g. not a some file in `extraSrcFiles`.

Again, I'm very unhappy to see this. You should have asked before accepting 10k diff.

Comment (phadej) [reply Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs:50]: In particular, there shouldn't be any *non-relative* files in `.cabal` files. The offenders like `extra-lib-dirs` simply don't belong into package files; those should be configured by users externally (e.g. in `cabal.project` or `cabal.config` files). That's the historical mistake which should be cleaned up, not further cemented.

Comment (geekosaur) [reply Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs:50]: Forcing users who want to install a package to turn it into a project or otherwise hack their config to make it work on e.g. MacOS isn't really acceptable.

Comment (phadej) [reply Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs:50]: I didn't mean you need to remove the support for `extra-lib-dirs` right away, but having system specific hard-coded paths is not correct either. The absolute paths do not belong into package definitions. That's why there are `build-type: Configure`, (soon `Hooks`) and `build-type: Custom`.

Comment (sheaf) [reply Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs:50]: > I'm very unhappy to see these kind of changes. You don't get any benefit. The `SymbolicPath PackageDir LicenseFile` was already _relative_, it's a (relative) path from package directory to the license file.
> 
> You also drop some type-safety, as `LicenseFile` was telling that it's not _some_ file, but a _license_ file (e.g. not a some file in `extraSrcFiles`.
> 
> Again, I'm very unhappy to see this. You should have asked before accepting 10k diff.

Oleg, I'm sorry but I spent a lot of time trying to come up with a design that was workable for the whole of Cabal. To do that, I did find it necessary to change the API in several ways. For instance, yes, we no longer specify the exact kind of file:

  - it lead to a proliferation of definitions for all the different kinds of files,
  - in my experience it was not useful to keep track of which specific file something was pointing to, it was extra book-keeping that didn't in practice avoid any bugs.
  
One other major point of departure came from the fact I realised we unfortunately would not be able to only use relative symbolic paths everywhere; there are a lot of parts in Cabal which use paths that are either relative or absolute, so I had to make that work with the API.

I understand if you think this "cemented" bad parts of the codebase, but I want to say that it is much easier now to change these things now because it is often a case of doing a search/replace, whereas when most of the codebase was "untyped" (= using `FilePath`) it took a lot of effort to identify the places which needed to be changed.

I know and can tell you put serious consideration into the design of the initial API. I honestly did try sticking to it as much as I could but I was also informed by looking at what was actually in use in the Cabal codebase. Maybe you think that's a mistake, and that I should have stuck to a design and tried harder to make Cabal adapt to that abstract ideal, but I don't think I would have managed to write that patch with that approach.

Comment (phadej) [reply Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs:50]: >  there are a lot of parts in Cabal which use paths 

That is the main pain point.

The `SymbolicPath` was intended to be used as *symbolic path*, i.e. paths to be interpreted that appear in `.cabal` files. Not any real filepath which may appear on the filesystem. That should had different type. `FilePath` or `OsPath` or ... I.e. not as typed wrapper around `filepath` / `directory`. Maybe it's not the case, but the distinction is now very blurry. (In particular, `CWD` root doesn't make sense for `SymbolicPath`s as paths in `.cabal` file in my design).

TL;DR I hope that `SymbolicPath` is not the only abstraction over paths, as in my opinion, it simply won't work well. Too many concerns, too complicated API.

File: Cabal/src/Distribution/Simple/Setup/Build.hs [hunk 4]
```diff
@@ -18,7 +21,15 @@
...
-  ( BuildFlags (..)
+  ( BuildFlags
+      ( BuildCommonFlags
+      , buildVerbosity
+      , buildDistPref
+      , buildCabalFilePath
+      , buildWorkingDir
+      , buildTargets
+      , ..
+      )
```
Comment (ulidtko) [inline Cabal/src/Distribution/Simple/Setup/Build.hs:32]: **Breaking change** to Cabal library users. :boom: 

```haskell
import Distribution.Simple.Setup (BuildFlags (buildVerbosity))

miniRepro :: BuildFlags
miniRepro = emptyBuildFlags {buildVerbosity = undefined}
```

— compiles against Cabal 3.12; yields compile error with 3.14:

```
src/Distribution/Extra/Doctest.hs:235:13: error: [GHC-16444]
    • non-bidirectional pattern synonym ‘Cabal-3.14.0.0:Distribution.Simple.Setup.Build.BuildCommonFlags’ used in an expression
    • In a record update at field ‘buildVerbosity’
      and with pattern synonym ‘Cabal-3.14.0.0:Distribution.Simple.Setup.Build.BuildCommonFlags’.
      In the expression: emptyBuildFlags {buildVerbosity = undefined}
      In an equation for ‘miniRepro’:
          miniRepro = emptyBuildFlags {buildVerbosity = undefined}
    |
235 | miniRepro = emptyBuildFlags {buildVerbosity = undefined}
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
